apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metricsdk-plugin/src/main/java/co/elastic/apm/agent/otelmetricsdk/MetricSetSerializer.java

```diff
@@ -18,53 +18,44 @@
  */
 package co.elastic.apm.agent.otelmetricsdk;
 
-import co.elastic.apm.agent.report.Reporter;
-import co.elastic.apm.agent.report.serialize.DslJsonSerializer;
-import com.dslplatform.json.BoolConverter;
-import com.dslplatform.json.DslJson;
-import com.dslplatform.json.JsonWriter;
-import com.dslplatform.json.Nullable;
-import com.dslplatform.json.NumberConverter;
+import co.elastic.apm.agent.tracer.reporting.DataWriter;
+import co.elastic.apm.agent.tracer.reporting.ReportingTracer;
 import io.opentelemetry.api.common.AttributeKey;
 import io.opentelemetry.api.common.AttributeType;
 import io.opentelemetry.api.common.Attributes;
 
+import javax.annotation.Nullable;
 import java.util.List;
 import java.util.Map;
 
-import static com.dslplatform.json.JsonWriter.ARRAY_END;
-import static com.dslplatform.json.JsonWriter.ARRAY_START;
-import static com.dslplatform.json.JsonWriter.COMMA;
-import static com.dslplatform.json.JsonWriter.OBJECT_END;
-import static com.dslplatform.json.JsonWriter.OBJECT_START;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.ARRAY_END;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.ARRAY_START;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.NEXT;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.NEW;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.OBJECT_END;
+import static co.elastic.apm.agent.tracer.reporting.DataWriter.StructureType.OBJECT_START;
 
 class MetricSetSerializer {
```

Comment:
Thanks for refactoring this duplicated json serialization logic, it has been there for a while but we never managed to get to it. Again, this allows to fix https://github.com/elastic/apm-agent-java/issues/3016